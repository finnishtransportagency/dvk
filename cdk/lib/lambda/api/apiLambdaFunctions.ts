import * as path from 'path';

interface BackendAPILambda {
  entry: string;
  pathPattern: string;
  functionName: string;
  priority: number;
  useVpc?: boolean;
  useMonitoring?: boolean;
}

const apiLambdaFunctions: BackendAPILambda[] = [
  {
    entry: path.join(__dirname, 'featureloader-handler.ts'),
    pathPattern: '/api/featureloader',
    functionName: 'featureloader',
    priority: 10,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'aislocations-handler.ts'),
    pathPattern: '/api/aislocations',
    functionName: 'aislocations',
    priority: 14,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'aisvessels-handler.ts'),
    pathPattern: '/api/aisvessels',
    functionName: 'aisvessels',
    priority: 16,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'pilotroute-handler.ts'),
    pathPattern: '/api/pilotroutes',
    functionName: 'pilotroutes',
    priority: 18,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'login.ts'),
    pathPattern: '/yllapito/kirjaudu.html',
    functionName: 'login',
    priority: 20,
    useVpc: false,
  },
  {
    entry: path.join(__dirname, 'logout.ts'),
    pathPattern: '/yllapito/api/logout',
    functionName: 'logout',
    priority: 30,
    useVpc: false,
  },
];

export default apiLambdaFunctions;
