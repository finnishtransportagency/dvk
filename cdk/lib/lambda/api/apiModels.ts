import { Geometry } from 'geojson';
import { GeometryPoint, Text } from '../../../graphql/generated';
import { FairwayCardIdName } from '../db/fairwayCardDBModel';

export type VaylaGeojsonFeature = { type?: string; geometry: Geometry };

/* VATU */
type MitoitusAlusAPIModel = {
  alustyyppiKoodi: string;
  alustyyppi: string;
  pituus: number;
  leveys: number;
  syvays: number;
  koko?: number;
  runkoTaytelaisyysKerroin?: number;
};

type LuokitusAPIModel = {
  luokitusTyyppi: string;
  vaylaluokkaKoodi: string;
  vaylaluokkaFI?: string;
  vaylaluokkaSV?: string;
};

export type VaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  vaylalajiKoodi?: string;
  vaylaLajiFI?: string;
  vaylaLajiSV?: string;
  valaistusKoodi?: string;
  valaistusFI?: string;
  valaistusSV?: string;
  omistaja?: string;
  merialueFI?: string;
  merialueSV?: string;
  alunSeloste?: string;
  paatepisteenSeloste?: string;
  normaaliKaantosade?: number;
  minimiKaantosade?: number;
  normaaliLeveys?: number;
  minimiLeveys?: number;
  varavesi?: string;
  lisatieto?: string;
  mareografi?: string;
  mitoitusalus?: MitoitusAlusAPIModel[];
  luokitus?: LuokitusAPIModel[];
};

export type VaylaFeature = { properties: VaylaAPIModel };
export type VaylaFeatureCollection = { type?: string; features: VaylaFeature[] };

type AlueVaylaAPIModel = {
  jnro: number;
  status: number;
  linjaus: number;
  mitoitusNopeus?: number;
  mitoitusNopeus2?: number;
  vaylaalueenJarjestysNro?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type AlueProperties = {
  id: number;
  nimi?: string;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  suunta?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  vayla?: AlueVaylaAPIModel[];
  tyyppiKoodi?: number;
  tyyppi?: string;
  merkintalajiKoodi?: number;
  merkintalaji?: string;
  liikennointiStatusKoodi?: number;
  liikennointiStatus?: string;
  liikennointiTyyppiKoodi?: number;
  liikenteenTyyppi?: string;
  liikennointiSuuntaKoodi?: number;
  liikennointiSuunta?: string;
};

export type AlueFeature = {
  properties: AlueProperties;
} & VaylaGeojsonFeature;

export type AlueFeatureCollection = {
  type?: string;
  features: AlueFeature[];
};

type NavigointiLinjaVaylaAPIModel = {
  jnro: number;
  status?: number;
  linjaus?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type NavigointiLinjaProperties = {
  id: number;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  tosisuunta?: number;
  pituus?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  tyyppiKoodi?: string;
  tyyppi?: string;
  vayla: NavigointiLinjaVaylaAPIModel[];
};

export type NavigointiLinjaFeature = {
  properties: NavigointiLinjaProperties;
} & VaylaGeojsonFeature;

export type NavigointiLinjaFeatureCollection = {
  type?: string;
  features: NavigointiLinjaFeature[];
};

type RajoitustyyppiAPIModel = {
  koodi?: string;
  rajoitustyyppi?: string;
};

type RajoitusVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

type RajoitusAlueProperties = {
  id: number;
  rajoitustyyppi?: string;
  rajoitustyypit?: RajoitustyyppiAPIModel[];
  suuruus?: number;
  esittaja?: string;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  muutosPaivamaara?: string;
  alkuPaivamaara?: string;
  loppuPaivamaara?: string;
  paatosTila?: string;
  tietolahde?: string;
  sijainti?: string;
  kunta?: string;
  poikkeus?: string;
  vayla?: RajoitusVaylaAPIModel[];
};

export type RajoitusAlueFeature = { properties: RajoitusAlueProperties } & VaylaGeojsonFeature;

export type RajoitusAlueFeatureCollection = {
  type?: string;
  features: RajoitusAlueFeature[];
};

export type TurvalaiteVaylaAPIModel = {
  jnro: number;
  paavayla?: string;
};

export type TurvalaiteVikatiedotProperties = {
  vikaId: number;
  turvalaiteNumero: number;
  turvalaiteNimiFI: string;
  turvalaiteNimiSV?: string;
  vikatyyppiKoodi: number;
  vikatyyppiFI: string;
  vikatyyppiSV?: string;
  vikatyyppiEN?: string;
  kirjausAika: string;
};
export type TurvalaiteVikatiedotFeature = { properties: TurvalaiteVikatiedotProperties } & VaylaGeojsonFeature;
export type TurvalaiteVikatiedotFeatureCollection = {
  type?: string;
  features: TurvalaiteVikatiedotFeature[];
};

type TurvalaiteReunaetaisyysAPIModel = {
  vaylaalueID: number;
  etaisyys: number;
};

type TurvalaiteProperties = {
  turvalaitenumero: number;
  nimiFI?: string;
  nimiSV?: string;
  alityyppi?: string;
  turvalaitetyyppiKoodi?: number;
  turvalaitetyyppiFI?: string;
  turvalaitetyyppiSV?: string;
  navigointilajiKoodi?: number;
  navigointilajiFI?: string;
  navigointilajiSV?: string;
  valaistu?: string;
  omistajaFI?: string;
  omistajaSV?: string;
  symboli?: string;
  kaukohallinta?: number;
  AISTyyppi?: number;
  AISTyyppiSeliteFI?: string;
  AISTyyppiSeliteSV?: string;
  vayla?: TurvalaiteVaylaAPIModel[];
  reunaetaisyys?: TurvalaiteReunaetaisyysAPIModel[];
};

export type TurvalaiteFeature = { properties: TurvalaiteProperties } & VaylaGeojsonFeature;

export type TurvalaiteFeatureCollection = {
  type?: string;
  features: TurvalaiteFeature[];
};

export type TaululinjaVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

export type TaululinjaProperties = {
  taululinjaId: number;
  suunta: number;
  etuTLId?: number;
  keskiTLId?: number;
  takaTLId?: number;
  vayla: TaululinjaVaylaAPIModel[];
};

export type TaululinjaFeature = { properties: TaululinjaProperties } & VaylaGeojsonFeature;
export type TaululinjaFeatureCollection = {
  type?: string;
  features: TaululinjaFeature[];
};

type KaantoympyraVaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  luokitus: number;
};

export type KaantoympyraProperties = {
  kaantoympyraID: number;
  halkaisija: number;
  haraussyvyys?: number;
  vertaustaso?: string;
  lisatieto?: string;
  vayla?: KaantoympyraVaylaAPIModel[];
};

export type KaantoympyraFeature = { properties: KaantoympyraProperties } & VaylaGeojsonFeature;

export type KaantoympyraFeatureCollection = { type?: string; features: KaantoympyraFeature[] };

/* Traficom */
export type PilotPlace = {
  id: number;
  name: Text;
  geometry: GeometryPoint;
  fairwayCards: FairwayCardIdName[];
};

/* Pooki */
export type MarineWarningDates = {
  startDateTime?: number;
  endDateTime?: number;
  dateTime?: number;
};

/* Weather */
export type WeatherMareograph = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WLEV_PT1S_INSTANT: number;
  WLEVN2K_PT1S_INSTANT: number;
};

export type Mareograph = {
  id: number;
  geometry: Geometry;
  name: string;
  dateTime: number;
  waterLevel: number;
  n2000WaterLevel: number;
  calculated: boolean;
};

export type WeatherObservation = {
  fmisid: number;
  geoid: number;
  latlon: string;
  stationname: string;
  localtime: string;
  WG_PT10M_MAX: number;
  WD_PT10M_AVG: number;
  WS_PT10M_AVG: number;
  TA_PT1M_AVG: number;
  VIS_PT1M_AVG: number | null;
};

export type Observation = {
  id: number;
  name: string;
  windSpeedAvg: number;
  windSpeedMax: number;
  windDirection: number;
  visibility: number | null;
  temperature: number;
  dateTime: number;
  geometry: Geometry;
};

export type WeatherBuoy = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WH_PT1M_ACC: number | null;
  WHD_PT1M_ACC: number | null;
  TW_PT1M_AVG: number | null;
};

export type Buoy = {
  id: number;
  name: string;
  dateTime: number;
  geometry: Geometry;
  temperature: number | null;
  waveDirection: number | null;
  waveHeight: number | null;
};

/* AIS */
export type VesselAPIModel = {
  name: string;
  timestamp: number;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number;
  eta: number;
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

type VesselLocationProperties = {
  mmsi: number;
  sog: number;
  cog: number;
  navStat: number;
  rot: number;
  posAcc: boolean;
  raim: boolean;
  heading?: number;
  timestamp: number;
  timestampExternal: number;
};

type VesselLocationFeature = {
  mmsi: number;
  type?: string;
  geometry: Geometry;
  properties: VesselLocationProperties;
};

export type VesselLocationFeatureCollection = {
  type?: string;
  dataUpdatedTime?: string;
  features: VesselLocationFeature[];
};

export type Vessel = {
  name: string;
  timestamp: Date;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number; // meters
  eta: string; // MM-DD HH:mm
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

/* RTZ */
export type Coordinate = [number, number];

export type RtzWaypoint = {
  coordinate: Coordinate;
  turnRadius: number;
};

type RtzGeometria = {
  type: string;
  coordinates: Coordinate;
};

export type RtzReittipiste = {
  tunnus: number;
  nimi: string;
  rtzTunniste: number;
  reittitunnus: number;
  kaarresade: number;
  geometria: RtzGeometria;
  leveysVasen: number;
  leveysOikea: number;
  geometriaTyyppi: string;
  muutosaikaleima: string;
  jarjestys: number;
};

export type RtzData = {
  tunnus: number;
  tila: number;
  nimi: string;
  tunniste: string;
  rtz: string;
  reittipisteet: Array<RtzReittipiste>;
};

/* IBNet */
export type IBNetApiResponse = {
  rvEndpoints: string[];
  toRv: string;
};

export type Dirway = {
  id: string;
  name: string;
  description: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type DirwayPoint = {
  id: string;
  dirway_id: string;
  order_num: number;
  name: string;
  latitude: number;
  longitude: number;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type LocationApiModel = {
  id: string;
  name: string;
  type: 'PORT' | 'FAIRWAY';
  locode_list: string;
  nationality: string;
  latitude: number;
  longitude: number;
  winterport: boolean;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type RestrictionApiModel = {
  id: string;
  location_id: string;
  start_time: string;
  end_time?: string | null;
  text_compilation: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};
