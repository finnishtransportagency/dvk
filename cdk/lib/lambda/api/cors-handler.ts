import { log } from '../logger';

export const handler = async (): Promise<object> => {
  log.debug(`cors()`);
  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': 'localhost',
      'Access-Control-Allow-Methods': '*',
    },
  };
};
