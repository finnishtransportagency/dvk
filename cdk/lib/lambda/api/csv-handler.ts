import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `csv()`);
  // TODO implement
  return {
    statusCode: 200,
    body: JSON.stringify({ hello: 'Hello from Lambda!' }),
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': 'http://localhost:3000',
      'Access-Control-Allow-Methods': '*',
      'Access-Control-Allow-Headers': '*',
    },
  };
};
