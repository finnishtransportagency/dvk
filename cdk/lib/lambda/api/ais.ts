import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Vessel, VesselAPIModel, VesselLocationFeatureCollection } from './apiModels';
import { fetchAISFeatureCollection, fetchAISMetadata } from './axios';

const aisV1Path = 'ais/api/ais/v1/';

function getTimestampParams(): Record<string, string> {
  const now = Date.now();
  const yesterday = Date.now() - 24 * 60 * 60 * 1000;
  return { from: yesterday.toString(), to: now.toString() };
}

function parseFrom64BitInteger(integer: number, bitStart: number, bitEnd: number) {
  // 64-bit integer: get bit range from start to end by losing other bits
  // Shift bits first to left and then to right, let others fall off
  const i = (integer << (63 - bitStart)) >>> (63 - bitStart + bitEnd);
  return i.toString().padStart(2, '0');
}

/* MMDDHHMM UTC
Bits 19-16: month; 1-12; 0 = not available = default
Bits 15-11: day; 1-31; 0 = not available = default
Bits 10-6: hour; 0-23; 24 = not available = default
Bits 5-0: minute; 0-59; 60 = not available = default
*/
function parseETA(eta: number): string {
  const month = parseFrom64BitInteger(eta, 19, 16);
  const day = parseFrom64BitInteger(eta, 15, 11);
  const hour = parseFrom64BitInteger(eta, 10, 6);
  const minute = parseFrom64BitInteger(eta, 5, 0);
  return month + '-' + day + ' ' + hour + ':' + minute;
}

export async function fetchVessels(): Promise<Vessel[]> {
  const path = aisV1Path + 'vessels';
  const params = getTimestampParams();
  const data: VesselAPIModel[] = await fetchAISMetadata(path, params);
  return data.map((row) => {
    return {
      name: row.name,
      timestamp: new Date(row.timestamp),
      mmsi: row.mmsi,
      callSign: row.callSign,
      imo: row.imo,
      shipType: row.shipType,
      draught: row.draught / 10,
      eta: parseETA(row.eta),
      posType: row.posType,
      referencePointA: row.referencePointA,
      referencePointB: row.referencePointB,
      referencePointC: row.referencePointC,
      referencePointD: row.referencePointD,
      destination: row.destination,
    };
  });
}

export async function fetchLocations(): Promise<FeatureCollection> {
  const path = aisV1Path + 'locations';
  const params = getTimestampParams();
  const data: VesselLocationFeatureCollection = await fetchAISFeatureCollection(path, params);
  const { dataUpdatedTime, features } = data;
  const geojsonFeatures: Feature<Geometry, GeoJsonProperties>[] = features.map((f) => {
    return {
      type: 'Feature',
      id: f.mmsi,
      geometry: f.geometry,
      properties: {
        ...f.properties,
        dataUpdatedTime,
        timestampExternal: new Date(f.properties.timestampExternal),
        featureType: 'aisvessel',
      },
    };
  });
  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: geojsonFeatures,
  };
  return collection;
}
