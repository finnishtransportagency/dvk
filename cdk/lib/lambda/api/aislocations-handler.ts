import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
