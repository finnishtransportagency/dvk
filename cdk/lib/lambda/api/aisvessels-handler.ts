import { ALBEvent, ALBResult } from 'aws-lambda';
import { getHeaders } from '../environment';
import { log } from '../logger';
import { getFromCache } from '../graphql/cache';
import { fetchVessels } from './ais';
import { handleLoaderError, saveResponseToS3 } from '../util';

function getKey() {
  return 'aisvessels';
}

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aisvessels()`);
  const key = getKey();
  let base64Response: string | undefined;
  let statusCode = 200;
  const response = await getFromCache(key);
  if (!response.expired && response.data) {
    base64Response = response.data;
  } else {
    try {
      const vessels = await fetchVessels();
      log.debug('ais vessels: %d', vessels.length);
      base64Response = await saveResponseToS3(vessels, key);
    } catch (e) {
      const errorResult = handleLoaderError(response, e);
      base64Response = errorResult.body;
      statusCode = errorResult.statusCode;
    }
  }
  return {
    statusCode,
    body: base64Response,
    isBase64Encoded: true,
    multiValueHeaders: {
      ...getHeaders(),
      'Content-Type': ['application/json'],
    },
  };
};
