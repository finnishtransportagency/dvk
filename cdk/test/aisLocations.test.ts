import { mockClient } from 'aws-sdk-client-mock';
import { handler } from '../lib/lambda/api/aislocations-handler';
import { mockAISALBEvent } from './mocks';
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { sdkStreamMixin } from '@aws-sdk/util-stream';
import locationsJson from './data/aislocations.json';
import { Readable } from 'stream';
import { gunzip, gzip } from 'zlib';
import assert from 'assert';
import { FeatureCollection } from 'geojson';

const s3Mock = mockClient(S3Client);
const path = 'aislocations';

jest.mock('../lib/lambda/environment', () => ({
  getFeatureCacheDurationHours: () => 2,
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHeaders: () => {},
}));

const locations: FeatureCollection = {
  type: 'FeatureCollection',
  features: [
    {
      type: 'Feature',
      id: 111222000,
      geometry: {
        type: 'Point',
        coordinates: [20.85169, 55.770832],
      },
      properties: {
        mmsi: 111222000,
        sog: 0.1,
        cog: 346.5,
        navStat: 1,
        rot: 4,
        posAcc: true,
        raim: true,
        heading: 79,
        timestamp: 59,
        timestampExternal: new Date(1659212938646),
        dataUpdatedTime: '2023-10-19T07:06:13Z',
        featureType: 'aisvessel',
      },
    },
    {
      type: 'Feature',
      id: 333444000,
      geometry: {
        type: 'Point',
        coordinates: [19.612433, 59.1802],
      },
      properties: {
        mmsi: 333444000,
        sog: 6.5,
        cog: 205.7,
        navStat: 3,
        rot: 0,
        posAcc: true,
        raim: false,
        heading: 205,
        timestamp: 28,
        timestampExternal: new Date(1638940711289),
        dataUpdatedTime: '2023-10-19T07:06:13Z',
        featureType: 'aisvessel',
      },
    },
    {
      type: 'Feature',
      id: 55666000,
      geometry: {
        type: 'Point',
        coordinates: [30.190985, 59.886713],
      },
      properties: {
        mmsi: 55666000,
        sog: 5.3,
        cog: 74.9,
        navStat: 0,
        rot: 0,
        posAcc: true,
        raim: true,
        heading: 76,
        timestamp: 45,
        timestampExternal: new Date(1539814847556),
        dataUpdatedTime: '2023-10-19T07:06:13Z',
        featureType: 'aisvessel',
      },
    },
  ],
};

async function parseResponse(body: string): Promise<FeatureCollection> {
  const response = new Promise<Error | Buffer>((resolve, reject) =>
    gunzip(Buffer.from(body, 'base64'), (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    })
  );
  return JSON.parse((await response).toString()) as FeatureCollection;
}

async function createCacheResponse(collectionJson: object) {
  const collection = JSON.stringify(collectionJson);
  const zippedString = new Promise<Error | Buffer>((resolve, reject) =>
    gzip(Buffer.from(collection), (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    })
  );
  const body = new Readable();
  body.push((await zippedString).toString('base64'));
  body.push(null);
  return body;
}

let throwError = false;
jest.mock('../lib/lambda/api/axios', () => ({
  fetchAISFeatureCollection: () => {
    if (throwError) {
      throw new Error('Fetching from AIS api failed');
    }
    return locations;
  },
}));

beforeEach(() => {
  jest.resetAllMocks();
  s3Mock.reset();
  throwError = false;
});

it('should get locations from api', async () => {
  const expires = new Date();
  expires.setTime(expires.getTime() + 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: sdkStreamMixin(await createCacheResponse(locationsJson)), Expires: expires });
  const response = await handler(mockAISALBEvent(path));
  assert(response.body);
  const responseObj = await parseResponse(response.body);
  expect(responseObj.features.length).toBe(3);
  expect(responseObj).toMatchSnapshot();
});

it('should get locations from cache when api call fails', async () => {
  const expires = new Date();
  expires.setTime(expires.getTime() - 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: sdkStreamMixin(await createCacheResponse(locationsJson)), Expires: expires });
  throwError = true;
  const response = await handler(mockAISALBEvent(path));
  assert(response.body);
  const responseObj = await parseResponse(response.body);
  expect(responseObj.features.length).toBe(4);
  expect(responseObj).toMatchSnapshot();
});

it('should get internal server error when api call fails and no cached response', async () => {
  throwError = true;
  const response = await handler(mockAISALBEvent(path));
  expect(response.statusCode).toBe(500);
});
