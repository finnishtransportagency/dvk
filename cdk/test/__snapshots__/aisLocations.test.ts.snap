// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get locations from api 1`] = `
{
  "features": [
    {
      "geometry": {
        "coordinates": [
          20.85169,
          55.770832,
        ],
        "type": "Point",
      },
      "properties": {
        "cog": 346.5,
        "heading": 79,
        "mmsi": 111222000,
        "navStat": 1,
        "posAcc": true,
        "raim": true,
        "rot": 4,
        "sog": 0.1,
        "timestamp": 59,
        "timestampExternal": "2022-07-30T20:28:58.646Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          19.612433,
          59.1802,
        ],
        "type": "Point",
      },
      "properties": {
        "cog": 205.7,
        "heading": 205,
        "mmsi": 333444000,
        "navStat": 3,
        "posAcc": true,
        "raim": false,
        "rot": 0,
        "sog": 6.5,
        "timestamp": 28,
        "timestampExternal": "2021-12-08T05:18:31.289Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          30.190985,
          59.886713,
        ],
        "type": "Point",
      },
      "properties": {
        "cog": 74.9,
        "heading": 76,
        "mmsi": 55666000,
        "navStat": 0,
        "posAcc": true,
        "raim": true,
        "rot": 0,
        "sog": 5.3,
        "timestamp": 45,
        "timestampExternal": "2018-10-17T22:20:47.556Z",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`should get locations from cache when api call fails 1`] = `
{
  "features": [
    {
      "geometry": {
        "coordinates": [
          21.164027,
          55.64427,
        ],
        "type": "Point",
      },
      "id": 777888000,
      "properties": {
        "cog": 4.6,
        "dataUpdatedTime": "2023-10-19T07:06:13Z",
        "heading": 324,
        "mmsi": 777888000,
        "navStat": 5,
        "posAcc": true,
        "raim": false,
        "rot": 0,
        "sog": 0,
        "timestamp": 49,
        "timestampExternal": "2023-05-27T20:14:52.125Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          19.95236,
          57.639723,
        ],
        "type": "Point",
      },
      "id": 999000000,
      "properties": {
        "cog": 203.5,
        "dataUpdatedTime": "2023-10-19T07:06:13Z",
        "heading": 204,
        "mmsi": 999000000,
        "navStat": 0,
        "posAcc": true,
        "raim": true,
        "rot": 0,
        "sog": 13,
        "timestamp": 34,
        "timestampExternal": "2023-04-09T22:53:35.445Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          18.679073,
          59.301408,
        ],
        "type": "Point",
      },
      "id": 445566000,
      "properties": {
        "cog": 10,
        "dataUpdatedTime": "2023-10-19T07:06:13Z",
        "heading": 511,
        "mmsi": 445566000,
        "navStat": 15,
        "posAcc": false,
        "raim": true,
        "rot": -128,
        "sog": 0,
        "timestamp": 25,
        "timestampExternal": "2023-05-01T02:40:24.648Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          20.18021,
          55.685625,
        ],
        "type": "Point",
      },
      "id": 778899000,
      "properties": {
        "cog": 95.4,
        "dataUpdatedTime": "2023-10-19T07:06:13Z",
        "heading": 95,
        "mmsi": 778899000,
        "navStat": 0,
        "posAcc": true,
        "raim": false,
        "rot": 0,
        "sog": 11.7,
        "timestamp": 49,
        "timestampExternal": "2022-08-13T14:28:55.921Z",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
